# Function 5 - Chemical Yield

### First Inspection

We are told, according to the function description provided, that the black-box function $f(x_1, x_2, x_3, x_4)$ represents the yield of a chemical process in a factory. Very importantly, we also told that this distribution in typically unimodal with a single peak where yield is maximised. Let us PCA for visualisation,

In [1]:
# Depedencies,
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA

# SKOPT imports,
from skopt import gp_minimize
from skopt.space import Real
from skopt.learning import GaussianProcessRegressor
from skopt.learning.gaussian_process.kernels import Matern, ConstantKernel, WhiteKernel
from skopt import Optimizer
from joblib import dump, load

# Loading known evaluations,
X, y = np.load("initial_inputs.npy"), np.load("initial_outputs.npy")

# Additional known evalutations,
X_add = np.array([
    [0.257096385587263, 0.915878528550171, 1, 1],
    [0.0688294279786485, 1, 1, 1],
    [0, 1, 1, 1],
    [0.0283191213741659, 0.863402556248803, 1, 1]
])

y_add = np.array([3486.87701870642, 4440.7345118934, 4440.5225, 3029.2624131587])

X = np.concatenate((X, X_add), axis=0)
y = np.concatenate((y, y_add), axis=0)

# Inspecting evalutation points,
X

array([[0.19144708, 0.03819337, 0.60741781, 0.41458414],
       [0.75865295, 0.53651774, 0.65600038, 0.36034155],
       [0.43834987, 0.8043397 , 0.21024527, 0.15129482],
       [0.70605083, 0.53419196, 0.26424335, 0.48208755],
       [0.83647799, 0.19360965, 0.6638927 , 0.78564888],
       [0.68343225, 0.11866264, 0.82904591, 0.56757661],
       [0.55362148, 0.66734998, 0.32380582, 0.81486975],
       [0.35235627, 0.32224153, 0.11697937, 0.47311252],
       [0.15378571, 0.72938169, 0.42259844, 0.44307417],
       [0.46344227, 0.63002451, 0.10790646, 0.9576439 ],
       [0.67749115, 0.35850951, 0.47959222, 0.07288048],
       [0.58397341, 0.14724265, 0.34809746, 0.42861465],
       [0.30688872, 0.31687813, 0.62263448, 0.09539906],
       [0.51114177, 0.817957  , 0.72871042, 0.11235362],
       [0.43893338, 0.77409176, 0.37816709, 0.93369621],
       [0.22418902, 0.84648049, 0.87948418, 0.87851568],
       [0.72526172, 0.47987049, 0.08894684, 0.75976022],
       [0.35548161, 0.63961937,

Despite the sparsity of data points, the domain of the black box function seems to be $[0, 1]^4$.

In [2]:
# Inspecting black-box outputs,
print(np.array2string(y, suppress_small=True))

[  64.44343986   18.3013796     0.1129398     4.21089813  258.37052545
   78.43438889   57.57153693  109.57187556    8.84799176  233.22361017
   24.42308831   64.42014682   63.47671579   79.72912993  355.80681776
 1088.8596182    28.86675164   45.18157035  431.61275676    9.97233189
 3486.87701871 4440.73451189 4440.5225     3029.26241316]


The chemical yield is positive and can vary by numerous magnitudes between each point. Therefore, we scale the function outputs,

In [3]:
def scale_yields(y, y_mean, y_std):
    y_scaled = (y - y_mean) / y_std
    return y_scaled

y_mean, y_std = np.mean(y), np.std(y)
y = scale_yields(y, y_mean, y_std)
print(np.array2string(y, suppress_small=True))

[-0.49569752 -0.52822501 -0.54104681 -0.53815799 -0.35899012 -0.48583471
 -0.50054182 -0.46388458 -0.5348891  -0.37671725 -0.52390956 -0.49571394
 -0.49637901 -0.48492199 -0.29030316  0.22645674 -0.52077703 -0.50927603
 -0.23686435 -0.53409651  1.91692055  2.58933501  2.58918555  1.59432867]


### Optimiser Configuration

From the discription of the black-box function, we know,

- $f(\mathbf{x})$ is likely to be unimodal and have a single global maximum. 
- The black-box function is not noisy (problem discription has not mentioned noise).

However, we do NOT know,

- If black-box function $f(\mathbf{x})$ is very smooth, generally smooth or rough.

As a result, our go-to Matérn 5/2 kernel (with ARD). Assuming that $f(\mathbf{x})$ is unimodal, we should consider exploitation instead of exploration. Therefore, we select PI as our acquistion function $\text{PI}(\mathbf{x})$. 

$$
\text{PI}(\mathbf{x}) = \Phi \left( \frac{\mu(\mathbf{x}) - f(\mathbf{x}^{+}) - \epsilon}{\sigma{\mathbf{x}}} \right)
$$

Our strategy is to slowly _"climb"_ the peak of the distribution by exploiting large known values of $f(\mathbf{x})$. To be pendantic, a given element of a chosen kernel is given by,

$$
K(\mathbf{x}_i, \mathbf{x}_j)
=
\sigma^2
\frac{2^{1-\nu}}{\Gamma(\nu)}
\left(
\sqrt{2\nu}\, r_{ij}
\right)^{\nu}
K_{\nu}
\left(
\sqrt{2\nu}\, r_{ij}
\right),
$$

$$
r_{ij}
=
\sqrt{
\frac{(x_{i1} - x_{j1})^2}{\ell_1^2}
+
\frac{(x_{i2} - x_{j2})^2}{\ell_2^2}
+
\frac{(x_{i3} - x_{j3})^2}{\ell_3^2}
+
\frac{(x_{i4} - x_{j4})^2}{\ell_4^2}
}.
$$

where we have set the lengthscales as $\ell_1 = \ell_2 = \ell_3 = \ell_4 = 0.1$ and used $\nu = 5/2$ as well as $\sigma = 1$.

### Optimiser Initialisation

Note that we had to re-initialise the optimiser on the $n=4$ iteration because the function yields were normalised.

In [4]:
"""INITALISING THE OPTIMISATION MODEL."""

# Inputting the given evaluations provided by the problem,
X_supplied = X.tolist()
y_supplied = y.tolist()

"""OPTIMISER SETTINGS."""

# We define the domain of the black-box function (or the range of the parameter values we want to consider),
space = [Real(0, 1, name="x1"),
         Real(0, 1, name="x2"),
         Real(0, 1, name="x3"),
         Real(0, 1, name="x4")
         ]

# Creating the kernel for the GPR,
kernel = ConstantKernel(1.0) * Matern(
    length_scale=(0.1, 0.1, 0.1, 0.1),
    length_scale_bounds=(1e-2, 1.0),
    nu = 5/2)

# GPR settings,
gpr = GaussianProcessRegressor(
    kernel=kernel,
    normalize_y=True,
    n_restarts_optimizer=10
)

# Creating optimisier,
opt = Optimizer(
    dimensions=space,
    base_estimator=gpr,
    acq_func="EI",
    acq_func_kwargs={"xi": 0.5},
    random_state=0
)

"""CREATING INTIAL OPTIMISER STATE."""

# Supplying given points to optimiser,
opt.tell(X_supplied, (-np.array(y_supplied)).tolist()) # <-- We flip the values since we are trying to maximise the black-box function.

# Asking for the next point to evaluate the black-box function,
point_query = opt.ask()

# Saving optimiser state (zero-th iteration),
dump(opt, "bayes_opt_state_iter4.joblib")

# Printing point query,
print(f"Point Query: {point_query}")

Point Query: [0.8604768471051507, 0.9970354477634344, 0.8851647973784308, 0.9553353487297673]


### Next Query 

In [64]:
current_query = 5

# Input the new evaluation,
X_new = [[]]
y_new = []

# Loading the previous state of the optimiser,
opt = load(f"bayes_opt_state_iter{current_query - 1}.joblib")

# Supplying the new query to the optimiser,
opt.tell(X_new, (-np.array(y_new)).tolist()) # <-- We flip the values since we are trying to maximise the black-box function.

# Asking for the next point to evaluate the black-box function,
point_query = opt.ask()

# Saving optimiser state,
dump(opt, f"bayes_opt_state_iter{current_query}.joblib")

# Printing point query,
print(f"Point Query {current_query + 1}: {point_query}")

ValueError: Not all points have the same dimensions as the space.

In [5]:
def create_query_string(query):
    """Convets the a given query into the query string format required for submission to the hosted black-box function."""

    # Concat,
    query_string = ""
    for coord in point_query:
        query_string += f"-{coord:.6f}"

    # Removing first character,
    query_string = query_string[1:]

    return query_string
    
print(create_query_string(point_query))

0.028319-0.863403-1.000000-1.000000


### Updates

Week 1: The optimiser sampled a point in close proximity to known point with the highest evaluation value. This is not unexpected as PI is considered to be the most greedy and exploitive of the acquisition functions available in `skopt`. However, it does appear to be too greedy in that we are not "climbing" the peak of the unimodal distribution quickly enough. For this reason, we replace PI with EI and restart the optimiser. We hope that EI remains primariliy exploitative, but is not as greedy such that we can "climb" towards global maximum at a faster pace.

Week 2: The evaluation from the previous week returned the highest chemical yield seen so far. From the generated new query, it seems as though the optimiser is exploring the boundary region of the space near the previous query.

Week 3: The chemical yield increases once more. As a result, the optimiser is choosing to sample a boundary point. If our optimiser remains stuck along the boundary of the space for the next few queries, we may have to change the value of $\xi$.

Week 4: It seems that the optimiser is stuck along the boundary of the search space. We will change the value of $\xi$ next week to encorage more exploration.

Week 5: The chemical yield is now scaled appropriately (the $n=4$ optimiser had to be re-initialised.). $\xi$ was also increased to $\xi=0.05$ to encorage more exploration. The generated query $\mathbf{x} = (0.860, 0.997, 0.885, 0.955)$ seems reasonable, it is along the boundary region near the highest yielding point so far $\mathbf{x} = (0.069, 1, 1, 1)$. The optimiser wants to explore the boundary regions since the more central regions have give low yields and most of the search space by volume is towards the edges.

Week 6: